plutus-contract-exe/src/Language/Plutus/Contract/Record.hs

```diff
@@ -0,0 +1,69 @@
+{-# LANGUAGE DeriveAnyClass     #-}
+{-# LANGUAGE DeriveGeneric      #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE LambdaCase         #-}
+module Language.Plutus.Contract.Record where
+
+import           Control.Lens
+import           Data.Bifunctor (Bifunctor (..))
+import           Data.Sequence  (Seq)
+import           GHC.Generics   (Generic)
+
+import           Data.Aeson     (Value)
+import qualified Data.Aeson     as Aeson
+
+-- | The serialisable state of a contract instance, containing a mix of raw 
+--   input events and serisalised checkpoints.
+--   See note [Handling state in contracts] in 'Language.Plutus.Contract.State'.
+type Record i = Either (OpenRecord i) (ClosedRecord i)
+
+data FinalValue i = FinalJSON Value | FinalEvents (Seq i)
+    deriving stock (Eq, Show, Generic, Functor)
+    deriving anyclass (Aeson.FromJSON, Aeson.ToJSON)
+
+data ClosedRecord i =
```

Comment:
I think what you may want to do here is something like:
```
-- Whichever language extensions you need to get type promotion
{-# LANGUAGE DataKinds #-}

type Closure = Open | Closed

data Record c i where
    RecordLeaf :: i -> Closure c i
    RecordBothClosed :: Record Closed i -> Record Closed i -> Record Closed i
    RecordOpenLeft :: Record Open i -> record Closed i -> Record Open i
    ...
```
Possibly you could do something clever with type families to avoid the four cases there, but I'm not clever enough for that.

This *should* mean that if you pattern match on a `Record Closed i` you'll only get the leaf and both closed cases.
    ... -- other binary closed/open cases

Reply:
After working through the GADT code that we started in our pairing session it turned out that we always need to know which of the two branches was closed and which one was open, so there is no place where having a `Record Open i` actually reduces the number of cases we need to look at. So I reverted back to the previous version.